llvm/lib/Analysis/ConstantFolding.cpp

```diff
@@ -2884,6 +2894,40 @@ static Constant *ConstantFoldIntrinsicCall2(
     const CallBase *Call) {
   assert(Operands.size() == 2 && "Wrong number of operands.");
 
+  // EVM local begin
+  if (Operands[0]->getType()->isIntegerTy() &&
+      Operands[1]->getType()->isIntegerTy()) {
+    const APInt *C0 = nullptr, *C1 = nullptr;
+    if (!getConstIntOrUndef(Operands[0], C0) ||
+        !getConstIntOrUndef(Operands[1], C1))
+      return nullptr;
+
+    if (IntrinsicID == Intrinsic::evm_signextend ||
+        IntrinsicID == Intrinsic::evm_exp ||
+        IntrinsicID == Intrinsic::evm_byte) {
+      if (isa<PoisonValue>(Operands[0]) || isa<PoisonValue>(Operands[1]) ||
+          !C0 || !C1)
+        return PoisonValue::get(Ty);
+
+      const unsigned BitWidth = Ty->getIntegerBitWidth();
+      assert(BitWidth == 256);
+
+      if (IntrinsicID == Intrinsic::evm_signextend) {
+        // Signextend operation returns original value if the extension
+        // overflows the type width.
+        if (C0->uge(APInt(BitWidth, (BitWidth / 8) - 1)))
```

Comment:
We perform this exact check on both call sites, shouldn't we move it to the callee?